### PR TITLE
optimize keyswitching

### PR DESCRIPTION
**Pull request by GelilaSeifu**
Avoids precomputation of NTT object for better performance when possible.

### CODE REVIEW

File: hexl/experimental/seal/key-switch-internal.cpp [hunk 1]
```diff
@@ -11,6 +11,7 @@
 #include "hexl/eltwise/eltwise-fma-mod.hpp"
 #include "hexl/eltwise/eltwise-mult-mod.hpp"
 #include "hexl/eltwise/eltwise-reduce-mod.hpp"
+#include "hexl/experimental/seal/get-ntt.hpp"
```
Comment (joserochh) [inline hexl/experimental/seal/key-switch-internal.cpp:14]: Can we change the name of this library to something like ntt-objs.hpp or ntt-cache.hpp?

File: hexl/experimental/seal/key-switch-internal.cpp [hunk 2]
```diff
@@ -11,6 +11,7 @@
...
 #include "hexl/logging/logging.hpp"
 #include "hexl/ntt/ntt.hpp"
```
Comment (joserochh) [inline hexl/experimental/seal/key-switch-internal.cpp:16]: Is this still needed after adding new get ntt library?

File: hexl/include/hexl/experimental/seal/get-ntt.hpp [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+struct HashPair {
+  template <class T1, class T2>
+  std::size_t operator()(const std::pair<T1, T2>& p) const {
+    auto hash1 = std::hash<T1>{}(p.first);
+    auto hash2 = std::hash<T2>{}(p.second);
+    return hash_combine(hash1, hash2);
+  }
+
+  static std::size_t hash_combine(std::size_t lhs, std::size_t rhs) {
+    lhs ^= rhs + 0x9e3779b9 + (lhs << 6) + (lhs >> 2);
```
Comment (joserochh) [inline hexl/include/hexl/experimental/seal/get-ntt.hpp:20]: Can we add a comment here saying this is based on Golden Ratio Hashing with seeds?

**Review by joserochh — CHANGES_REQUESTED** [review]

**Review by joserochh — APPROVED** [review]
LG

**Review by jlhcrawford — APPROVED** [review]

**Pull request merged**